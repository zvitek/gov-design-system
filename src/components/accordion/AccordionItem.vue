<template>
    <div>
        <button
            class="gov-accordion__header"
            :id="uniqueId"
            :aria-expanded="isOpen"
            @click="toggle"
            :class="{
                'is-expanded': isOpen,
                'gov-accordion__header--noborder': noBorder}">
            <h3>
                <slot name="icon"/>
                {{ label }}
            </h3>
        </button>
        <div
            class="gov-accordion__content"
            :aria-labelledby="uniqueId"
            v-show="isOpen"
            tabindex="-1"
            :class="{
            'is-expanded': isOpen}">
            <div class="gov-accordion__content-inner">
                <slot/>
            </div>
        </div>
    </div>
</template>

<script>
import { makeid } from '../../utils/helpers'

export default {
    name: 'GovAccordionItem',
    model: {
        prop: 'open',
        event: 'update:open'
    },
    props: {
        open: {
            type: Boolean,
            default: false
        },
        ariaId: {
            type: String,
            required: false,
            default: null
        },
        label: {
            type: String,
            required: true
        },
        noBorder: {
            type: Boolean,
            required: false
        }
    },
    data() {
        return {
            isOpen: this.open,
            uniqueId: null
        }
    },
    watch: {
        open(value) {
            this.isOpen = value
        }
    },
    methods: {
        /**
         * Toggle and emit events
         */
        toggle() {
            this.isOpen = !this.isOpen
            this.$emit('update:open', this.isOpen)
            this.$emit(this.isOpen ? 'open' : 'close')
        }
    },
    mounted() {
        this.uniqueId = this.ariaId || 'accordion_' + makeid()
    }
}
</script>
