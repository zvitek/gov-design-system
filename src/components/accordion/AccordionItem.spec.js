import { shallowMount } from '@vue/test-utils'
import AccordionItem from '@components/accordion/AccordionItem'

let wrapper

describe('AccordionItem', () => {
    describe('default', () => {
        beforeEach(() => {
            wrapper = shallowMount(AccordionItem, {
                propsData: {
                    open: true,
                    label: 'Acc Label'
                }
            })
        })

        it('is called', () => {
            expect(wrapper.name()).toBe('GovAccordionItem')
            expect(wrapper.isVueInstance()).toBeTruthy()
        })

        it('default props and vm', () => {
            expect(wrapper.props().open).toBe(true)
            expect(wrapper.vm.isOpen).toBe(true)
        })

        it('render correctly', () => {
            expect(wrapper.html()).toMatchSnapshot()
        })
    })
    describe('open prop is false', () => {
        beforeEach(() => {
            wrapper = shallowMount(AccordionItem, {
                propsData: {
                    open: false,
                    label: 'Acc Label'
                }
            })
        })

        it('set default open prop', () => {
            expect(wrapper.props().open).toBe(false)
            expect(wrapper.vm.isOpen).toBe(false)
        })

        it('call toggle method', async () => {
            expect(wrapper.find('.gov-accordion__content').isVisible()).toBe(false)
            wrapper.vm.toggle()
            await wrapper.vm.$nextTick()
            expect(wrapper.vm.isOpen).toBe(true)
            expect(wrapper.find('.gov-accordion__content').isVisible()).toBe(true)
        })

        it('emit a click event', () => {
            expect(wrapper.find('.gov-accordion__content').isVisible()).toBe(false)
            const updateOpen = jest.fn()
            wrapper.vm.$on('update:open', updateOpen)
            wrapper.find('.gov-accordion__header').trigger('click')
            expect(updateOpen).toHaveBeenCalledTimes(1)
            expect(updateOpen).toHaveBeenCalledWith(true)
            expect(wrapper.vm.isOpen).toBe(true)
            expect(wrapper.find('.gov-accordion__content').isVisible()).toBe(true)
        })

        it('should emit open event', () => {
            const open = jest.fn()
            wrapper.vm.$on('open', open)
            wrapper.find('.gov-accordion__header').trigger('click')
            expect(open).toHaveBeenCalledTimes(1)
        })

        it('update open prop', () => {
            expect(wrapper.find('.gov-accordion__content').isVisible()).toBe(false)
            wrapper.setProps({ open: true })
            expect(wrapper.vm.isOpen).toBe(true)
            expect(wrapper.find('.gov-accordion__content').isVisible()).toBe(true)
        })
    })

    it('should have icon slot', () => {
        const iconSlot = '<strong> Icon </strong>'
        const wrapper = shallowMount(AccordionItem, {
            scopedSlots: {
                icon: iconSlot
            }
        })
        expect(wrapper.find('.gov-accordion__header h3 :first-child').html()).toBe(iconSlot)
    })

    it('should have default slot', () => {
        const slotDefault = '<div> Content </div>'
        wrapper = shallowMount(AccordionItem, {
            slots: {
                default: slotDefault
            }
        })
        expect(wrapper.find('.gov-accordion__content-inner :first-child').html()).toBe(slotDefault)
    })

    it('should emit close event', () => {
        wrapper = shallowMount(AccordionItem, {
            propsData: {
                open: true
            }
        })
        const close = jest.fn()
        wrapper.vm.$on('close', close)
        wrapper.find('.gov-accordion__header').trigger('click')
        expect(close).toHaveBeenCalledTimes(1)
    })
})
