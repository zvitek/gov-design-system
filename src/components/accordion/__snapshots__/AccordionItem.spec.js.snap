// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItem default render correctly 1`] = `
<div><button id="" aria-expanded="true" class="gov-accordion__header is-expanded">
        <h3>

        </h3>
    </button>
    <div aria-labelledby="" tabindex="-1" class="gov-accordion__content is-expanded">
        <div class="gov-accordion__content-inner"></div>
    </div>
</div>
`;
